prereise/gather/demanddata/nrel_efs/get_efs_data.py

```diff
@@ -36,129 +33,273 @@ def download_data(es=None, ta=None, fpath=""):
+        ta = {"Slow", "Moderate", "Rapid"}
+
+    # Check that the components of ta are valid
     if not ta.issubset({"Slow", "Moderate", "Rapid"}):
         invalid_ta = ta - {"Slow", "Moderate", "Rapid"}
         raise ValueError(f'Invalid electrification scenarios: {", ".join(invalid_ta)}')
 
+    # Return the reformatted ta
+    return ta
+
+
+def _check_path(fpath):
+    """Checks the file path input to :py:func:`download_demand_data`,
+    :py:func:`download_flexibility_data`, :py:func:`partition_demand_by_sector`, and
+    :py:func:`partition_flexibility_by_sector`.
+
+    :param str fpath: The input file path.
+    :return: (*str*) -- The necessary file path in case it needed to be accessed.
+    :raises TypeError: if fpath is not input as a str.
+    """
+
+    # Check that the input is of an appropriate type
+    if not isinstance(fpath, str):
+        raise TypeError("The file path must be input as a str.")
+
     # Access the actual path if not already provided
     if len(fpath) == 0:
         fpath = os.getcwd()
 
-    # Download each of the specified load profiles
-    z = {}
-    for i in es:
-        z[i] = {}
-        for j in ta:
-            # Assign path and file names
-            zip_name = "EFSLoadProfile_" + i + "_" + j + ".zip"
-            url = "https://data.nrel.gov/system/files/126/" + zip_name
-            zip_path = os.path.join(fpath, zip_name)
+    # Return fpath in case it had to be accessed
+    return fpath
 
-            # Save a local copy of the .zip file for extraction
-            r = requests.get(url, stream=True)
-            if r.status_code != requests.codes.ok:
-                r.raise_for_status()
-            open(zip_name, "wb").write(r.content)
-            print(zip_name + " successfully downloaded!")
 
-            # Store the data in memory to try extracting with Python's zipfile module
-            z[i][j] = zipfile.ZipFile(io.BytesIO(r.content))
+def _download_data(zip_name, url, fpath):
+    """Downloads the specified NREL EFS data for :py:func:`download_demand_data` and
+    :py:func:`download_flexibility_data`.
 
-    # Try to extract the .csv file from the .zip file
-    zf_works = True
-    for i in es:
-        for j in ta:
-            # Assign path and file names
-            zip_name = "EFSLoadProfile_" + i + "_" + j + ".zip"
-            csv_name = "EFSLoadProfile_" + i + "_" + j + ".csv"
-            zip_path = os.path.join(fpath, zip_name)
+    :param str zip_name: The name of the specified .zip file.
+    :param str url: The specified URL to access the desired .zip file.
+    :param str fpath: The input file path.
+    :return: (*zipfile.ZipFile*) -- The .zip file stored in memory for attempted
+        extraction using Python's zipfile module.
+    """
+
+    # Save a local copy of the .zip file for extraction
+    r = requests.get(url, stream=True)
+    if r.status_code != requests.codes.ok:
+        r.raise_for_status()
+    open(zip_name, "wb").write(r.content)
```

Comment:
I think it is generally cleaner to do `with open(...) as f: f.write(...)`

Reply:
I can certainly change `open(zip_name, "wb").write(r.content)` if that is preferred. However, I don't really understand how this is cleaner. @danielolsen, can you explain?

Reply:
@lanesmith218 Using `with` statement will automatically close the file and clean the cache after interactions. It is a fancier way to do `try: open...finally: close` provided by Python3.

Reply:
Thanks for clarifying! I'll make that change.

Reply:
In general, classes can have `__enter__` and `__exit__` methods, and the `with` syntax ensures that `__enter__` is done before anything in the block, and `__exit__` is done as a `finally` step (i.e. always executed). For `open` specifically, this ensures that the file is closed afterwards.

Reply:
Done.